leveldb_broker.go

```diff
@@ -0,0 +1,185 @@
+package gocelery
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"sync"
+
+	"github.com/syndtr/goleveldb/leveldb"
+)
+
+// levelDBBroker implements CeleryBroker backed by levelDB
+type levelDBBroker struct {
+	db    *leveldb.DB
+	queue string
+	head  uint
+	tail  uint
+
+	mu sync.Mutex // protect head and tail
+}
+
+// levelDBBrokerState is used to store the Queue state in levelDB
+type levelDBBrokerState struct {
+	Head uint `json:"head"`
+	Tail uint `json:"tail"`
+}
+
+// NewLevelDBBroker returns levelDB backed implementation of CeleryBroker
+// We will also initialise any previous state
+func NewLevelDBBroker(db *leveldb.DB, queue string) CeleryBroker {
+	brk := &levelDBBroker{
+		db:    db,
+		queue: queue,
+		head:  0,
+		tail:  0,
+	}
+
+	// restore the old state from the DB
+	brk.loadBrokerState()
+
+	return brk
+}
+
+// loadBrokerState loads any previous state if any available
+// logs any state retrieval errors
+func (lbrk *levelDBBroker) loadBrokerState() {
+	key := []byte(fmt.Sprintf("celery-leveldb-broker-state-%s", lbrk.queue))
+	data, err := lbrk.db.Get(key, nil)
+	if err != nil {
+		return
+	}
+
+	state := new(levelDBBrokerState)
+	err = json.Unmarshal(data, state)
+	if err != nil {
+		log.Printf("failed to load the previous state from levelDB: %v\n", err)
+		return
+	}
+
+	lbrk.head = state.Head
+	lbrk.tail = state.Tail
+}
+
+// saveBrokerState saves the broker state to the disk
+// due to limitation of worker, we shall save state on every update
+// to ensure the state is upto date on restart
+func (lbrk *levelDBBroker) saveBrokerState() error {
+	state := levelDBBrokerState{Head: lbrk.head, Tail: lbrk.tail}
+	data, err := json.Marshal(state)
+	if err != nil {
+		return err
+	}
+
+	key := []byte(fmt.Sprintf("celery-leveldb-broker-state-%s", lbrk.queue))
+	err = lbrk.db.Put(key, data, nil)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// count returns the current items in the queue
+func (lbrk *levelDBBroker) count() uint {
+	c := lbrk.tail - lbrk.head
+	if c < 0 {
+		c = 0
+	}
+
+	return c
+}
+
+// GetTaskMessage returns the first message in the levelDB queue
+func (lbrk *levelDBBroker) GetTaskMessage() (*TaskMessage, error) {
+	lbrk.mu.Lock()
+	defer lbrk.mu.Unlock()
+
+	if lbrk.count() == 0 {
+		return nil, nil
+	}
+
+	// get the task
+	key := []byte(fmt.Sprintf("celery-leveldb-task-%v", lbrk.head))
```

Comment:
Looks like here we are using prefix `celery-leveldb-task-` but we prefix in the DB with `celery-leveldb-result-` is that intended as a compound key? otherwise the `lbrk.head` would be enough.

Reply:
`lbrk.head` would not be enough since anyone can override that unintentionally.   Each backend/brokers will have its own prefixes so that re-use is not easily possible unless intentional

Reply:
Why the broker does not use the backend interface instead that will already prefix the key lookup/write? Or the broker uses a different prefix than the backend?

Reply:
Backend/broker interface doesn't do that. It is up to the implementation of each interface that should handle it. Redis uses it own prefixes so that it doesn't clash with any other keys.

Reply:
Got it, separate storage for backend and broker. No need further discussion.